### PR TITLE
feat: add support for internal functions directory

### PR DESCRIPTION
**Pull request by eduardoboucas**
**- Summary**

Adds support for bundling functions from `.netlify/functions-internal`. Needs https://github.com/netlify/build/pull/3213.

**- Test plan**

Added new tests.

**- A picture of a cute animal (not mandatory but encouraged)**

5bd39fe9929a7-wallpaper-preview

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -75,7 +75,7 @@
     "prettier": "--ignore-path .gitignore --loglevel=warn \"{src,scripts,site,tests,.github}/**/*.{js,md,yml,json,html}\" \"*.{js,yml,json,html}\" \".*.{js,yml,json,html}\" \"!CHANGELOG.md\" \"!npm-shrinkwrap.json\""
   },
   "dependencies": {
-    "@netlify/build": "^15.11.5",
+    "@netlify/build": "^16.0.0-rc",
```
Comment (eduardoboucas) [inline package.json:78]: To be changed after https://github.com/netlify/build/pull/3213 is released.

Comment (eduardoboucas) [reply package.json]: Reverted.

**Comment by eduardoboucas** [comment]
With https://github.com/netlify/cli/pull/2975, this is now ready for review.

File: src/lib/functions/registry.js [hunk 2]
```diff
@@ -102,18 +102,7 @@ class FunctionsRegistry {
...
-    }
-
-    this.functions.set(name, func)
-    this.buildFunctionAndWatchFiles(func)
-
-    this.logger.log(`${NETLIFYDEVLOG} ${chalk.green('Loaded')} function ${chalk.yellow(name)}.`)
-  }
-
-  async scan(directory) {
+  async prepareDirectoryScan(directory) {
```
Comment (eduardoboucas) [inline src/lib/functions/registry.js:105]: `prepareDirectoryScan` is now responsible for creating the directory and running any `onDirectoryScan` hooks. It runs at the beginning of `scan` for every configured directory.

File: src/lib/functions/registry.js [hunk 3]
```diff
@@ -128,8 +117,29 @@ class FunctionsRegistry {
...
+    }
+
+    this.functions.set(name, func)
+    this.buildFunctionAndWatchFiles(func)
+
+    this.logger.log(`${NETLIFYDEVLOG} ${chalk.green('Loaded')} function ${chalk.yellow(name)}.`)
+  }
+
+  async scan(directory) {
+    const directories = (Array.isArray(directory) ? directory : [directory]).filter(Boolean)
```
Comment (erezrokah) [inline src/lib/functions/registry.js:134]: [dust] Could we always pass an array to simplify the code?

File: src/utils/functions/index.js [hunk 4]
```diff
@@ -1,8 +1,23 @@
...
   config.functionsDirectory ||
   (config.dev && config.dev.Functions) ||
   defaultValue
 
-module.exports = { getFunctionsDir }
+const getInternalFunctionsDir = async () => {
+  const path = getPathInProject(['functions-internal'])
+
+  try {
+    const stat = await statAsync(path)
```
Comment (erezrokah) [inline src/utils/functions/index.js:15]: [sand] As we already have it as a dependency, can we use https://github.com/sindresorhus/path-type#isdirectorypath?

File: tests/serving-functions.test.js [hunk 5]
```diff
@@ -445,6 +445,128 @@ export { handler }
...
+          }),
+        })
+        .buildAsync()
+
+      await withDevServer({ cwd: builder.directory, args }, async ({ port }) => {
+        t.is(await got(`http://localhost:${port}/.netlify/functions/hello`).text(), 'User')
+
+        await pause(WAIT_WRITE)
+
+        await builder
```
Comment (erezrokah) [inline tests/serving-functions.test.js:537]: [sand] Is this part of the test required?
Looks like we're already checking that `hello` return `'User'` before.

File: tests/command.deploy.test.js [hunk 6]
```diff
@@ -398,6 +398,63 @@ if (process.env.NETLIFY_TEST_DISABLE_LIVE !== 'true') {
...
+      const { deploy_url: deployUrl } = await callCli(
+        ['deploy', '--build', '--json'],
+        {
+          cwd: builder.directory,
+          env: { NETLIFY_SITE_ID: t.context.siteId },
+        },
+        true,
+      )
+
+      t.is(await got(`${deployUrl}/.netlify/functions/func-1`).text(), 'User 1')
```
Comment (erezrokah) [inline tests/command.deploy.test.js:451]: [sand] Do we want to test that user provider functions override internal ones in the deploy?
Is that the expected behavior?

**Review by erezrokah — APPROVED** [review]
This looks 🚀 

I've added some comments, non blocking

File: tests/serving-functions.test.js (hunk 5 above)
Comment (eduardoboucas) [reply tests/serving-functions.test.js:537]: We're updating both the user and the internal functions and asserting that the user function continues to take precedence.

File: tests/command.deploy.test.js (hunk 6 above)
Comment (eduardoboucas) [reply tests/command.deploy.test.js:451]: We are testing that. We upload two functions named `test-2` (https://github.com/netlify/cli/pull/2954/files/3c72fb4d0f63465fde7e7222e613efdda5939466#diff-f4cfe11a0a7c59e35d96895571eecb4c3386c79f7258d6a6b9222efa6010643aR418 and https://github.com/netlify/cli/pull/2954/files/3c72fb4d0f63465fde7e7222e613efdda5939466#diff-f4cfe11a0a7c59e35d96895571eecb4c3386c79f7258d6a6b9222efa6010643aR425) and asserting that the user-defined one takes precedence.

Comment (erezrokah) [reply tests/command.deploy.test.js:451]: Oh, sorry I missed - please ignore my comment.

File: src/lib/functions/registry.js (hunk 3 above)
Comment (eduardoboucas) [reply src/lib/functions/registry.js]: Good point! Done in https://github.com/netlify/cli/pull/2954/commits/b9a89c6fc70a43771d1553e344ee9bf009f804b4.

File: src/utils/functions/index.js (hunk 4 above)
Comment (eduardoboucas) [reply src/utils/functions/index.js]: Done in https://github.com/netlify/cli/pull/2954/commits/036573c34f19868818b94a884f24dbc8cd60d5ce.

File: src/lib/functions/registry.js [hunk 7]
```diff
@@ -128,8 +117,27 @@ class FunctionsRegistry {
...
+    }
+
+    this.functions.set(name, func)
+    this.buildFunctionAndWatchFiles(func)
+
+    this.logger.log(`${NETLIFYDEVLOG} ${chalk.green('Loaded')} function ${chalk.yellow(name)}.`)
+  }
+
+  async scan(directories) {
+    if (directories.length === 0) {
```
Comment (erezrokah) [inline src/lib/functions/registry.js:134]: I think we still need to filter `null` directories as `getInternalFunctionsDir` can return `null` and it is used in https://github.com/netlify/cli/pull/2954/files#diff-de106b743be68ae2e9146d09931f894e258783a68db916a6d0e39e3908b28c84R165

Comment (eduardoboucas) [reply src/lib/functions/registry.js:134]: Good catch. Done in https://github.com/netlify/cli/pull/2954/commits/d1eb8ca56a6ba0890feda9e5773e3e41fa684538.

**Review by erezrokah — APPROVED** [review]
🚀

**Pull request merged**